### PR TITLE
Master

### LINKED ISSUE #72: Getting it to work in windows
Dont know if it's my Python knowledge that is a bit limited but I cant get it to work under windows.
Have done both "pip install jottalib" and "pip install jottacloudclient" with no errors but can't get it to load afterwards. is there something more that needs to be done?

/A

### LINKED ISSUE #73: Verify checksum after download
Via private email, the suggestion was raised that our tools could automatically verify the content after download, by comparing the md5 checksum from jottacloud and that of the newly downloaded file.

Sounds like a nice command line option for `download()` in `cli.py`. 

This would be a nice way to get the know the codebase for beginners.

### LINKED ISSUE #78: Support for folder download in Jotta-download
I am not sure, but it seems that jotta-download doesn't support download of whole folders.  There might be something else causing this error, below is the error I get.

```
Traceback (most recent call last):
  File "C:\Python27\Scripts\jotta-download-script.py", line 9, in <module>
    load_entry_point('jottalib==0.4.post1', 'console_scripts', 'jotta-download')()
  File "c:\python27\lib\site-packages\jottalib\cli.py", line 226, in download
    total_size = remote_file.size
AttributeError: 'JFSFolder' object has no attribute 'size'
```

### LINKED ISSUE #85: Don't use utf-8 in console output on windows
When running the jotta-dowload there seems to be an issue with the progress bar.
The error I get is:

```
C:\Users\XX>jotta-download Jottacloud.pdf
Traceback (most recent call last):
  File "C:\Python27\Scripts\jotta-download-script.py", line 9, in <module>
    load_entry_point('jottalib==0.4.1.post1', 'console_scripts', 'jotta-download')()
  File "c:\python27\lib\site-packages\jottalib\cli.py", line 229, in download
    with ProgressBar(expected_size=total_size) as bar:
  File "c:\python27\lib\site-packages\clint\textui\progress.py", line 63, in __init__
    self.show(0)
  File "c:\python27\lib\site-packages\clint\textui\progress.py", line 87, in show
    self.expected_size, self.etadisp))
  File "c:\python27\lib\site-packages\clint\packages\colorama\ansitowin32.py", line 34, in write
    self.__convertor.write(text)
  File "c:\python27\lib\site-packages\clint\packages\colorama\ansitowin32.py", line 115, in write
    self.write_and_convert(text)
  File "c:\python27\lib\site-packages\clint\packages\colorama\ansitowin32.py", line 142, in write_and_convert
    self.write_plain_text(text, cursor, len(text))
  File "c:\python27\lib\site-packages\clint\packages\colorama\ansitowin32.py", line 147, in write_plain_text
    self.wrapped.write(text[start:end])
  File "c:\python27\lib\encodings\cp850.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_map)
UnicodeEncodeError: 'charmap' codec can't encode characters in position 1-32: character maps to <undefined>
```

Commenting out line 229 and 233 It works perfectly.

### LINKED ISSUE #86: Downloading files and folders with Swedish letters fails
There seem to be some problem downloading files and folders that have Swedish letters (åäö).
The problem seem to be in urllib.py and in posixpath.py but is there something we can do in jottalib to prevent this to happend?

```
WARNING:py.warnings:c:\python27\lib\urllib.py:1303: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  return ''.join(map(quoter, s))

Traceback (most recent call last):
  File "C:\Python27\Scripts\jotta-download-script.py", line 9, in <module>
    load_entry_point('jottalib==0.4.1.post1', 'console_scripts', 'jotta-download')()
  File "c:\python27\lib\site-packages\jottalib\cli.py", line 273, in download
    remote_object = jfs.getObject(abs_path_to_object)
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 851, in getObject
    o = self.get(url)
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 838, in get
    url = self.escapeUrl(url)
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 807, in escapeUrl
    url = quote(urlparts[0], safe=self.rootpath)
  File "c:\python27\lib\urllib.py", line 1303, in quote
    return ''.join(map(quoter, s))
KeyError: u'\xe5'
```

```
jotta-download -c -v åäö.pdf
Traceback (most recent call last):
  File "C:\Python27\Scripts\jotta-download-script.py", line 9, in <module>
    load_entry_point('jottalib==0.4.1.post1', 'console_scripts', 'jotta-download')()
  File "c:\python27\lib\site-packages\jottalib\cli.py", line 226, in download
    path_to_object = posixpath.join(root_folder.path, args.remoteobject)
  File "c:\python27\lib\posixpath.py", line 73, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe5 in position 1: ordinal not in range(128)
```

### LINKED ISSUE #87: Parsing big XML files with "lxml.objectify.fromstring" returns an error
This is mainly related to #78 where an xml- file can grow quite big (in my case its around 500 mb and contains 779917 files and 90361 folders). But I guess this could happen otherwise too.

Anyway, there is an option to use a custom parser with the option "huge_tree" (http://stackoverflow.com/questions/11850345/using-python-lxml-etree-for-huge-xml-files). Would this be an option or is there another way of parsing large xml-files, for example in chunks?

```
reply: 'HTTP/1.1 200 OK\r\n'
header: Date: Wed, 27 Jan 2016 14:13:59 GMT
header: Accept-Ranges: bytes
header: path_list_total_files: 779917
header: path_list_total_folders: 90361
header: Content-Type: text/xml
header: Transfer-Encoding: chunked
header: Server: Jetty(8.1.4.v20120524)
DEBUG:requests.packages.urllib3.connectionpool:"GET /jfs/XX/Jotta/Sync/Backup2?mode=list HTTP/1.1" 200 None
Traceback (most recent call last):
  File "C:\Python27\Scripts\jotta-download-script.py", line 9, in <module>
    load_entry_point('jottalib==0.4.1.post1', 'console_scripts', 'jotta-download')()
  File "c:\python27\lib\site-packages\jottalib\cli.py", line 258, in download
    fileTree = remote_object.filedirlist().tree #Download the folder tree
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 304, in filedirlist
    return self.jfs.getObject(url)
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 851, in getObject
    o = self.get(url)
  File "c:\python27\lib\site-packages\jottalib\JFS.py", line 839, in get
    o = lxml.objectify.fromstring(self.raw(url))
  File "src/lxml/lxml.objectify.pyx", line 1801, in lxml.objectify.fromstring (src\lxml\lxml.objectify.c:26755)
  File "src/lxml/lxml.etree.pyx", line 3213, in lxml.etree.fromstring (src\lxml\lxml.etree.c:82934)
  File "src/lxml/parser.pxi", line 1819, in lxml.etree._parseMemoryDocument (src\lxml\lxml.etree.c:124533)
  File "src/lxml/parser.pxi", line 1707, in lxml.etree._parseDoc (src\lxml\lxml.etree.c:123074)
  File "src/lxml/parser.pxi", line 1079, in lxml.etree._BaseParser._parseDoc (src\lxml\lxml.etree.c:117114)
  File "src/lxml/parser.pxi", line 573, in lxml.etree._ParserContext._handleParseResultDoc (src\lxml\lxml.etree.c:110510)
  File "src/lxml/parser.pxi", line 683, in lxml.etree._handleParseResult (src\lxml\lxml.etree.c:112276)
  File "src/lxml/parser.pxi", line 624, in lxml.etree._raiseParseError (src\lxml\lxml.etree.c:111367)
lxml.etree.XMLSyntaxError: None
```

### LINKED ISSUE #88: remove md5 hash from incomplete files in JFSFileDirList()
When going through the code in the JFSFileDirList class i found that it takes into consideration that an incomplete file doesn't have a size. This is the case for md5 aswell and therefore I think the code should be:

```
                    else:
                        # an incomplete file
                        t.append(treefile(unicode(file_.attrib['name']),
                                          -1, # incomplete files have no size
                                          unicode(""), # incomplete files have no md5
                                          unicode(file_.attrib['uuid'])
                                          )
                                 )
            self.tree[posixpath.join(path, foldername)] = t
```
insted of the code:
```
                    else:
                        # an incomplete file
                        t.append(treefile(unicode(file_.attrib['name']),
                                          -1, # incomplete files have no size
                                          unicode(file_.currentRevision.md5),
                                          unicode(file_.attrib['uuid'])
                                          )
                                 )
            self.tree[posixpath.join(path, foldername)] = t
```

### PR DESCRIPTION
**Pull request by antonhagg**
Added functionality for folder download.
Added functionality for hash checking when downloading. 
Resolve #72
Resolve #73
Resolve #78
Resolve #85
Resolve #86
Resolve #87
Resolve #88

### CODE REVIEW

**Comment by havardgulldahl** [comment]
Ok, @antonhagg, thanks for all your work. I'll take a look.

File: src/jottalib/JFS.py [hunk 1]
```diff
@@ -839,9 +839,20 @@ def raw(self, url, extra_headers=None):
...
     def get(self, url):
-        'Make a GET request for url and return the response content as a generic lxml object'
+        'Make a GET request for url and return the response content as a generic lxml object'    
+        url = url.encode(encoding='UTF-8')
         url = self.escapeUrl(url)
-        o = lxml.objectify.fromstring(self.raw(url))
+        if "?mode=list" in url: #Check if we are requested a full tree of the directory
+            if os.path.exists('temp.xml'):
+                os.remove('temp.xml') 
+            with open("temp.xml", "w") as text_file:
```
Comment (havardgulldahl) [inline src/jottalib/JFS.py]: Could you update this to use tempfile.NamedTemporaryFile() instead?

Comment (antonhagg) [inline src/jottalib/JFS.py]: Should be possible. It has been good to have the xml for debugging.

**Comment by havardgulldahl** [comment]
So, @antonhagg, I have gone through the code now feature by feature, and added everything and rewrote parts of it. 

May I suggest next time you keep your commits to one feature at a time? It makes it much easier to follow both how your patch evolved, and to pull in your stuff.

But don't take this to hard. I really appreciate your time. Please keep contributing, your efforts are really helpful 👍

**Comment by antonhagg** [comment]
Wow, cool! =)
Since this is the first time I gave a shot at Python, i kinda expectet that it would need some modification and cleanup.
Also, I see the problem with adding all features in one bulk can make it hard to track individual things. I kinda lost overview myself after a while, so I'm impressed that you had a go at it. 

I appriciate the feedback, in still learing. =)

**Pull request closed without merging**